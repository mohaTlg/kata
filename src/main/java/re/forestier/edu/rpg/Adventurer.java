package re.forestier.edu.rpg;

import java.util.ArrayList;
import java.util.HashMap;

public class Adventurer extends Player {

    public Adventurer(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, money, inventory);
        setAbilities();
    }

    @Override
    public void setAbilities() {
        // Initialize abilities
        this.abilities = abilitiesPerTypeAndLevel().get(1);
    }

    @Override
    public String getAvatarClass() {
        return "ADVENTURER";
    }



    public HashMap<String, Integer> getAvatarLevel(int level){
        return abilitiesPerTypeAndLevel().get(level);
    }

    public HashMap<Integer, HashMap<String, Integer>> abilitiesPerTypeAndLevel() {
        HashMap<Integer, HashMap<String, Integer>> adventurerMap = new HashMap<>();

        Object[][] levels = {
                {1, new String[]{"INT", "DEF", "ATK", "CHA"}, new int[]{1, 1, 3, 2}},
                {2, new String[]{"INT", "CHA"}, new int[]{2, 3}},
                {3, new String[]{"ATK", "ALC"}, new int[]{5, 1}},
                {4, new String[]{"DEF"}, new int[]{3}},
                {5, new String[]{"VIS", "DEF"}, new int[]{1, 4}}
        };

        for (Object[] levelData : levels) {
            int level = (int) levelData[0];
            String[] abilities = (String[]) levelData[1];
            int[] values = (int[]) levelData[2];

            HashMap<String, Integer> levelAbilities = new HashMap<>();
            for (int i = 0; i < abilities.length; i++) {
                levelAbilities.put(abilities[i], values[i]);
            }
            adventurerMap.put(level, levelAbilities);
        }

        return adventurerMap;
    }

}
