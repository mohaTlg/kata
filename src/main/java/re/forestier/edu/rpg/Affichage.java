package re.forestier.edu.rpg;

import java.util.List;

public class Affichage {

    public static String afficherJoueur(Player player) {
        final String[] finalString = {"Joueur " + player.Avatar_name + " joué par " + player.playerName};
        finalString[0] += "\nNiveau : " + player.retrieveLevel() + " (XP totale : " + player.xp + ")";
        finalString[0] += "\n\nCapacités :";
        player.abilities.forEach((name, level) -> {
            finalString[0] += "\n   " + name + " : " + level;
        });
        finalString[0] += "\n\nInventaire :";
        player.inventory.forEach(item -> {
            finalString[0] += "\n   " + item;
        });

        return finalString[0];
    }

    public static String afficherJoeurMarkdown(Player player) {
        StringBuilder markdown = new StringBuilder();

        markdown.append("# Joueur : ").append(player.playerName).append("\n");
        markdown.append("## Avatar : ").append(player.Avatar_name).append("\n");
        markdown.append("**Classe d'avatar** : ").append(player.getAvatarClass()).append("\n");
        markdown.append("**Niveau** : ").append(player.retrieveLevel()).append("\n\n");
        markdown.append("### Capacités :\n");
        player.abilities.forEach((name, level) -> {
            markdown.append("* **").append(name).append("** : ").append(level).append("\n");
        });
        
        markdown.append("\n### Inventaire** :\n");
        player.inventory.forEach(item -> {
            markdown.append(" *").append(item.getName()).append("\n");
        });
        return markdown.toString();
    }


}
