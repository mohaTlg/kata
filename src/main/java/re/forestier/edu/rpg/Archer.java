package re.forestier.edu.rpg;

import java.util.ArrayList;
import java.util.HashMap;

public class Archer extends Player {

    public Archer(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, money, inventory);
        setAbilities();
    }

    @Override
    public void setAbilities() {
        this.abilities = abilitiesPerTypeAndLevel().get(1);
    }
    
    @Override
    public String getAvatarClass() {
        return "ARCHER";
    }

    public HashMap<String, Integer> getAvatarLevel(int level){
        return abilitiesPerTypeAndLevel().get(level);
    }

    public static HashMap<Integer, HashMap<String, Integer>> abilitiesPerTypeAndLevel() {
        HashMap<Integer, HashMap<String, Integer>> archerMap = new HashMap<>();

        Object[][] levels = {
                {1, new String[]{"INT", "ATK", "CHA", "VIS"}, new int[]{1, 3, 1, 3}},
                {2, new String[]{"DEF", "CHA"}, new int[]{1, 2}},
                {3, new String[]{"ATK"}, new int[]{3}},
                {4, new String[]{"DEF"}, new int[]{2}},
                {5, new String[]{"ATK"}, new int[]{4}}
        };

        for (Object[] levelData : levels) {
            int level = (int) levelData[0];
            String[] abilities = (String[]) levelData[1];
            int[] values = (int[]) levelData[2];

            HashMap<String, Integer> levelAbilities = new HashMap<>();
            for (int i = 0; i < abilities.length; i++) {
                levelAbilities.put(abilities[i], values[i]);
            }
            archerMap.put(level, levelAbilities);
        }

        return archerMap;
    }

}
