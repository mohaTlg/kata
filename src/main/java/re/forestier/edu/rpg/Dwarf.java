package re.forestier.edu.rpg;


import java.util.ArrayList;
import java.util.HashMap;

public class Dwarf extends Player {


    public Dwarf(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, money, inventory);
        setAbilities();
    } 

    @Override
    public void setAbilities() {
        this.abilities = abilitiesPerTypeAndLevel().get(1);
    }

    @Override
    public String getAvatarClass() {
        return "DWARF";
    }

    public HashMap<String, Integer> getAvatarLevel(int level){
        return abilitiesPerTypeAndLevel().get(level);
    }

    public static HashMap<Integer, HashMap<String, Integer>> abilitiesPerTypeAndLevel() {
        HashMap<Integer, HashMap<String, Integer>> dwarfMap = new HashMap<>();

        Object[][] levels = {
                {1, new String[]{"ALC", "INT", "ATK"}, new int[]{4, 1, 3}},
                {2, new String[]{"DEF", "ALC"}, new int[]{1, 5}},
                {3, new String[]{"ATK"}, new int[]{4}},
                {4, new String[]{"DEF"}, new int[]{2}},
                {5, new String[]{"CHA"}, new int[]{1}}
        };

        for (Object[] levelData : levels) {
            int level = (int) levelData[0];
            String[] abilities = (String[]) levelData[1];
            int[] values = (int[]) levelData[2];

            HashMap<String, Integer> levelAbilities = new HashMap<>();
            for (int i = 0; i < abilities.length; i++) {
                levelAbilities.put(abilities[i], values[i]);
            }
            dwarfMap.put(level, levelAbilities);
        }

        return dwarfMap;
    }

}
