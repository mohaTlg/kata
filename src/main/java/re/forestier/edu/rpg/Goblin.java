package re.forestier.edu.rpg;
import java.util.ArrayList;
import java.util.HashMap;

public class Goblin extends Player {

    public Goblin(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, money, inventory);
        setAbilities();
    }

    @Override
    public void setAbilities() {
        // Initialisation des capacités spécifiques au Gobelin
        this.abilities = abilitiesPerTypeAndLevel().get(1);
    }

    @Override
    public String getAvatarClass() {
        return "GOBLIN";
    }

    @Override
    public HashMap<String, Integer> getAvatarLevel(int level) {
        return abilitiesPerTypeAndLevel().get(level);
    }

    public HashMap<Integer, HashMap<String, Integer>> abilitiesPerTypeAndLevel() {
        HashMap<Integer, HashMap<String, Integer>> goblinMap = new HashMap<>();

        Object[][] levels = {
                {1, new String[]{"INT", "ATK", "ALC"}, new int[]{2, 2, 1}},
                {2, new String[]{"ATK", "ALC"}, new int[]{3, 4}},
                {3, new String[]{"VIS"}, new int[]{1}},
                {4, new String[]{"DEF"}, new int[]{1}},
                {5, new String[]{"DEF", "ATK"}, new int[]{2, 4}}
        };

        // Construction automatique
        for (Object[] levelData : levels) {
            int level = (int) levelData[0];
            String[] abilities = (String[]) levelData[1];
            int[] values = (int[]) levelData[2];

            HashMap<String, Integer> levelAbilities = new HashMap<>();
            for (int i = 0; i < abilities.length; i++) {
                levelAbilities.put(abilities[i], values[i]);
            }
            goblinMap.put(level, levelAbilities);
        }

        return goblinMap;
    }
}

