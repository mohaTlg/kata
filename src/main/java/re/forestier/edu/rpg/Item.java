package re.forestier.edu.rpg;

public class Item {
    private String name;
    private String description;
    private int weight;
    private int value;

    public Item(String name, String description, int weight, int value) {
        this.name = name;
        this.description = description;
        this.weight = weight;
        this.value = value;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public int getWeight() {
        return weight;
    }
    

    public int getValue() {
        return value;
    }

    @Override
    public String toString() {
        return name + " (" + description + ") - Poids: " + weight + ", Valeur: " + value;
    }
}
