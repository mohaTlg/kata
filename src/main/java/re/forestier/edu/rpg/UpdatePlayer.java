package re.forestier.edu.rpg;

import java.util.Map;
import java.util.HashMap;
import java.util.Random;

public class UpdatePlayer {

    private final static String[] objectList = {
        "Lookout Ring : Prevents surprise attacks",
        "Scroll of Stupidity : INT-2 when applied to an enemy", 
        "Draupnir : Increases XP gained by 100%", 
        "Magic Charm : Magic +10 for 5 rounds", 
        "Rune Staff of Curse : May burn your ennemies... Or yourself. Who knows?", 
        "Combat Edge : Well, that's an edge", 
        "Holy Elixir : Recover your HP"
    };

    public static HashMap<String, HashMap<Integer, HashMap<String, Integer>>> abilitiesPerTypeAndLevel() {
        HashMap<String, HashMap<Integer, HashMap<String, Integer>>> abilitiesPerTypeAndLevel = new HashMap<>();
    
        abilitiesPerTypeAndLevel.put("ADVENTURER", createAdventurerAbilities());
        abilitiesPerTypeAndLevel.put("ARCHER", createArcherAbilities());
        abilitiesPerTypeAndLevel.put("DWARF", createDwarfAbilities());
    
        return abilitiesPerTypeAndLevel;
    }
    
    private static HashMap<Integer, HashMap<String, Integer>> createAdventurerAbilities() {
        HashMap<Integer, HashMap<String, Integer>> adventurerMap = new HashMap<>();
        adventurerMap.put(1, createAbilitiesMap(new String[]{"INT", "DEF", "ATK", "CHA"}, new int[]{1, 1, 3, 2}));
        adventurerMap.put(2, createAbilitiesMap(new String[]{"INT", "CHA"}, new int[]{2, 3}));
        adventurerMap.put(3, createAbilitiesMap(new String[]{"ATK", "ALC"}, new int[]{5, 1}));
        adventurerMap.put(4, createAbilitiesMap(new String[]{"DEF"}, new int[]{3}));
        adventurerMap.put(5, createAbilitiesMap(new String[]{"VIS", "DEF"}, new int[]{1, 4}));
        return adventurerMap;
    }
    
    private static HashMap<Integer, HashMap<String, Integer>> createArcherAbilities() {
        HashMap<Integer, HashMap<String, Integer>> archerMap = new HashMap<>();
        archerMap.put(1, createAbilitiesMap(new String[]{"INT", "ATK", "CHA", "VIS"}, new int[]{1, 3, 1, 3}));
        archerMap.put(2, createAbilitiesMap(new String[]{"DEF", "CHA"}, new int[]{1, 2}));
        archerMap.put(3, createAbilitiesMap(new String[]{"ATK"}, new int[]{3}));
        archerMap.put(4, createAbilitiesMap(new String[]{"DEF"}, new int[]{2}));
        archerMap.put(5, createAbilitiesMap(new String[]{"ATK"}, new int[]{4}));
        return archerMap;
    }
    
    private static HashMap<Integer, HashMap<String, Integer>> createDwarfAbilities() {
        HashMap<Integer, HashMap<String, Integer>> dwarfMap = new HashMap<>();
        dwarfMap.put(1, createAbilitiesMap(new String[]{"ALC", "INT", "ATK"}, new int[]{4, 1, 3}));
        dwarfMap.put(2, createAbilitiesMap(new String[]{"DEF", "ALC"}, new int[]{1, 5}));
        dwarfMap.put(3, createAbilitiesMap(new String[]{"ATK"}, new int[]{4}));
        dwarfMap.put(4, createAbilitiesMap(new String[]{"DEF"}, new int[]{2}));
        dwarfMap.put(5, createAbilitiesMap(new String[]{"CHA"}, new int[]{1}));
        return dwarfMap;
    }
    
    private static HashMap<String, Integer> createAbilitiesMap(String[] abilities, int[] values) {
        HashMap<String, Integer> map = new HashMap<>();
        for (int i = 0; i < abilities.length; i++) {
            map.put(abilities[i], values[i]);
        }
        return map;
    }
    

    public static boolean addXp(player player, int xp) {
        int previousLevel = player.retrieveLevel();
        player.xp += xp;
        int newLevel = player.retrieveLevel();

        if (newLevel > previousLevel) { // Player leveled-up!

            // Give a random object
            Random random = new Random();
            player.inventory.add(objectList[random.nextInt(objectList.length)]);

            // Add/upgrade abilities to player
            Map<String, Integer> abilities = abilitiesPerTypeAndLevel().get(player.getAvatarClass()).get(newLevel);
            player.abilities.putAll(abilities);
            
            return true;
        }
        return false;
    }

    // majFinDeTour met Ã  jour les points de vie
    public static void majFinDeTour(player player) {
        if(player.currenthealthpoints == 0) {
            System.out.println("Le joueur est KO !");
            return;
        }

        if(player.currenthealthpoints < player.healthpoints/2) {
            if(!player.getAvatarClass().equals("ADVENTURER")) {
                if(player.getAvatarClass().equals("DWARF")) {
                    if(player.inventory.contains("Holy Elixir")) {
                        player.currenthealthpoints+=1;
                    }
                    player.currenthealthpoints+=1;
                } else if(player.getAvatarClass().equals("ADVENTURER")) {
                    player.currenthealthpoints+=2;
                }


                if(player.getAvatarClass().equals("ARCHER")) {
                    player.currenthealthpoints+=1;
                    if(player.inventory.contains("Magic Bow")) {
                        player.currenthealthpoints+=player.currenthealthpoints/8-1;
                    }
                }
            } else {
                player.currenthealthpoints+=2;
                if(player.retrieveLevel() < 3) {
                    player.currenthealthpoints-=1;
                }
            }
        } else if(player.currenthealthpoints >= player.healthpoints/2){
            if(player.currenthealthpoints >= player.healthpoints) {
                player.currenthealthpoints = player.healthpoints;
                return;
            }
        }


        if(player.currenthealthpoints >= player.healthpoints) {
            player.currenthealthpoints = player.healthpoints;
        }
    }
}