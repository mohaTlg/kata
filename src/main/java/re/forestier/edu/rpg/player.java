package re.forestier.edu.rpg;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;


public abstract class Player {

    public String playerName;
    public String Avatar_name;
    private String AvatarClass;

    public Integer money;
    private Float __real_money__;
    public int maxWeight;
    public int currentWeight;


    public int level;
    public int healthpoints;
    public int currenthealthpoints;
    protected int xp;


    public HashMap<String, Integer> abilities;
    public ArrayList<Item> inventory;  

    public Player(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        this.playerName = playerName;
        this.Avatar_name = avatarName;
        this.money = money;
        this.inventory = inventory;
    }

    public abstract void setAbilities();
    public abstract String getAvatarClass();
    public void setInventory(ArrayList<Item> inventory) {
        this.inventory = inventory;
    }


    public void removeMoney(int amount) throws IllegalArgumentException {
        if (money - amount < 0) {
            throw new IllegalArgumentException("Player can't have a negative money!");
        }
        money = money - amount;
    }
    public void addMoney(int amount) {
        var value = Integer.valueOf(amount);
        money = money + (value != null ? value : 0);
    }
    public int retrieveLevel() {
        HashMap<Integer, Integer> levels = new HashMap<>();
        levels.put(2,10);
        levels.put(3,27);
        levels.put(4,57);
        levels.put(5,111);
        //TODO : ajouter les prochains niveaux

        if (xp < levels.get(2)) {
            return 1;
        }
        else if (xp < levels.get(3)) {return 2;
        }
        if (xp < levels.get(4)) {
            return 3;
        }
        if (xp < levels.get(5)) return 4;
        return 5;
    }

    public int getXp() {
        return this.xp;
    }

    public void setXp(int xp) {
        this.xp = xp;
    }

    private final static Item[] objectList = {
            new Item("Lookout Ring", "Prevents surprise attacks", 2, 100),
            new Item("Scroll of Stupidity", "INT-2 when applied to an enemy", 1, 50),
            new Item("Draupnir", "Increases XP gained by 100%", 3, 300),
            new Item("Magic Charm", "Magic +10 for 5 rounds", 1, 75),
            new Item("Rune Staff of Curse", "May burn your enemies... Or yourself. Who knows?", 4, 150),
            new Item("Holy Elixir", "Recover your HP", 1, 200),
    };

    public abstract HashMap<String, Integer> getAvatarLevel(int level);

    public static boolean addXp(Player player, int xp) {
        int currentLevel = player.retrieveLevel();
        player.xp += xp;
        int newLevel = player.retrieveLevel();

        if (newLevel > currentLevel) {
            Random random = new Random();

            Item newItem = objectList[random.nextInt(objectList.length)];
            player.inventory.add(newItem);

            HashMap<String, Integer> abilities = player.getAvatarLevel(newLevel);
            abilities.forEach(player.abilities::put);

            return true;
        }
        return false;
    }


    // majFinDeTour met à jour les points de vie
    public static void majFinDeTour(Player player) {
        if (player.currenthealthpoints == 0) {
            System.out.println("Le joueur est KO !");
            return;
        }

        if (player.currenthealthpoints >= player.healthpoints) {
            player.currenthealthpoints = player.healthpoints;
            return;
        }

        if (player.currenthealthpoints < player.healthpoints / 2) {
            setHealthByClass(player);
        }
    }

    private static void setHealthByClass(Player player) {
        switch (player.getAvatarClass()) {
            case "DWARF":
                dwarHealth(player);
                break;

            case "ARCHER":
                archerHealth(player);
                break;

            case "ADVENTURER":
                adventurerHealth(player);
                break;

            default:
                break;
        }
    }

    private static void dwarHealth(Player player) {
        player.currenthealthpoints += 1;
    
        for (Item item : player.inventory) {
            if (item.getName().equals("Holy Elixir")) {
                player.currenthealthpoints += 1; 
                break;
            }
        }
    }
    

    private static void archerHealth(Player player) {
        player.currenthealthpoints += 1;
    
        for (Item item : player.inventory) {
            if (item.getName().equals("Magic Bow")) {
                player.currenthealthpoints += player.currenthealthpoints / 8 - 1;
                break; 
            }
        }
    }
    

    private static void adventurerHealth(Player player) {
        player.currenthealthpoints += 2;
        if (player.retrieveLevel() < 3) {
            player.currenthealthpoints -= 1;
        }
    }

    public boolean addItem(Item item) {
        if (currentWeight + item.getWeight() <= maxWeight) {
            inventory.add(item);                
            currentWeight += item.getWeight(); 
            System.out.println(item.getName() + " a été ajouter dans l'inventaire.");
            return true;
        } else {
            System.out.println("Vous ne pouvez pas porter plus d'objet !");
            return false;
        }
    }

}