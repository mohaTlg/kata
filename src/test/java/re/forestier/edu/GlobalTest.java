package re.forestier.edu;

import org.junit.jupiter.api.Test;

import re.forestier.edu.rpg.Adventurer;
import re.forestier.edu.rpg.Affichage;
import re.forestier.edu.rpg.Archer;
import re.forestier.edu.rpg.Item;
import re.forestier.edu.rpg.Player;

import java.util.ArrayList;

import static org.approvaltests.Approvals.verify;
import static org.junit.jupiter.api.Assertions.assertEquals;
public class GlobalTest {

    @Test
    void testAffichageBase() {
        Adventurer player = new Adventurer("Florian", "Gnognak le Barbare", 200, new ArrayList<Item>());
        Adventurer.addXp(player, 20);
        player.setInventory(new ArrayList<>());

        verify(Affichage.afficherJoueur(player));
    }

       @Test
    void testAffichageMarkdown() {
        Archer player = new Archer("Florian", "SpiderMan", 150, new ArrayList<>());

        verify(Affichage.afficherJoeurMarkdown(player));
    }
}
