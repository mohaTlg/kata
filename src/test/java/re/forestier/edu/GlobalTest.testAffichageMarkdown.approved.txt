# Joueur : Florian
## Avatar : SpiderMan
**Classe d'avatar** : ARCHER
**Niveau** : 1

### Capacit√©s :
* **VIS** : 3
* **ATK** : 3
* **CHA** : 1
* **INT** : 1

### Inventaire** :